R/GroupSequentialDesign.R

```diff
@@ -0,0 +1,183 @@
+#'     of the Gaussian quadrature rule to use for integration
+#'
+#' @rdname GSDesign-class
+#' @export
+GS_GQDesign <- function(n1, c1f, c1e, n2, c2_pivots, order) {
+    if (length(c2_pivots) != order)
+        stop("length of pivot vectors does not fit")
+    rule <- GaussLegendreRule(order)
+    GSDesign(n1 = n1, c1f = c1f, c1e = c1e, n2 = n2, c2_pivots = c2_pivots,
+             x1_norm_pivots = rule$nodes, weights = rule$weights)
+}
+
+
+
+#' @param x object to get parameters from
+#'
+#' @rdname GSDesign-class
+#' @export
+setMethod("tunable_parameters", signature("GSDesign"),
+          function(x, ...) c(x@n1, x@c1f, x@c1e, x@n2, x@c2_pivots))
+
+
+
+
+#' @param params vector of design parameters (must be in same order as returned
+#'     by \code{as.numeric(design)})
+#' @param object object to update
+#'
+#' @rdname GSDesign-class
+#' @export
+setMethod("update", signature("GSDesign"),
+    function(object, params, ...) {
+        k <- length(object@weights)
+        if( (length(params) - 4) != k)
+            stop("parameter length does not fit")
+        new("GSDesign",
+            n1  = params[1],
+            c1f = params[2],
+            c1e = params[3],
+            n2  = params[4],
+            c2_pivots = params[5:(length(params))],
+            x1_norm_pivots = object@x1_norm_pivots,
+            weights = object@weights)
+    })
+
+
+
+#' @param x1 stage-one outcome
+#' @param d design object
+#'
+#' @rdname GSDesign-class
+#' @export
+setMethod("n2", signature("GSDesign", "numeric"),
+          function(d, x1, ...) ifelse(x1 < d@c1f | x1 > d@c1e, 0, d@n2) )
+
+
+
+
+#' Convert a group-sequential design to a two-stage design
+#'
+#' @param d object of class \code{GSDesign}
+#'
+#' @export
+
+gs2ts <- function(d){
+    if(class(d) != "GSDesign")
+        stop("d must be of class GSDesign")
+    new("TwoStageDesign", n1 = d@n1, c1f = d@c1f, c1e = d@c1e,
+                    n2_pivots = rep(d@n2, length(d@weights)),  c2_pivots = d@c2_pivots,
+                    x1_norm_pivots = d@x1_norm_pivots, weights = d@weights)
+}
+
+
+
+#' Find optimal group-sequential design by constraint minimization
+#'
+#' \code{minimize} takes an unconditioonal score [TODO: need joint superclass for UnconditionalDesign and AffineScores...]
+#' and a constraint set (or single constraint) and solves the corresponding
+#' constraint minimization problem using \code{nloptr} (using COBYLA by default).
+#'
+#' @param objective objective function
+#' @param subject_to constraint collection
+#' @param initial_design initial guess (x0 for nloptr)
+#' @param lower_boundary_design design specifying the lower boundary
+#' @param upper_boundary_design design specifying the upper boundary
+#' @param opts options list passed to nloptr
+#' @param ... further optional arguments passed to \code{\link{nloptr}}
+#'
+#' @export
+minimize_gs <- function(objective, subject_to, initial_design,
```

Comment:
that is not acceptable! our minimize function will be universal ;) 

have a look at my implementation - minimize essentially calls tunable_parameters anyway. So it should work out of the box with ANY kind of design that you implement. The only thing that you need to do is to introduce a joint abstract superclass for GSDesign and TwoStageDesign because the generic function minimize() currently only has a method for TwoStageDesigns ;)

Reply:
As a general guideline: as soon as you copy-paste something there is probably a better way of doing it ;)

Reply:
also the name that you gave the function should make you suspicious. _gs means that this is the implementation for GSDesign objects, right? The whole point of OO is to make this explicit need to pick the right function go away by figuring out which method to call from the arguments that you pass!